### PR TITLE
Improve `__setattr__` performance of Pydantic models by caching setter functions

### PR DESCRIPTION
**Pull request by MarkusSintonen**
## Change Summary

Attribute setting has been pretty slow for `BaseModel` due to the extensive checks it has been doing for every `__setattr__` call. PR improves performance of `__setattr__` by memoizing the attribute specific handlers to the model class. This makes the attribute assigning some 7x faster. Also add missing benchmarks for attribute usage.

```python
from timeit import timeit
from pydantic import BaseModel

class Model(BaseModel):
    field: int

model = Model(field=1)

def run():
    model.field = 2

# Before 1.048
# After 0.147
print(timeit(run, number=1000000))
```

## Related issue number

fix https://github.com/pydantic/pydantic/issues/10853

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Review by sydney-runkle — COMMENTED** [review]
Hi @MarkusSintonen,

Cool idea, thanks! I think memoization could be helpful here. Let me circle back with some colleagues to verify.

Specifically, @dmontagu, wdyt about this? I recall you've done a lot of work on these setattr branches.

File: pydantic/main.py
Comment (dmontagu) [inline pydantic/main.py:173]: I think it's worth a bit more strength/detail in this comment, e.g.:
```suggestion
    """__setattr__ handlers. Memoizing the setattr handlers leads to a dramatic performance improvement in `__setattr__`"""
```
(that suggestion would probably need reformatting.)

File: pydantic/main.py
Comment (dmontagu) [reply pydantic/main.py:897]: I think the word `fast` in the name of this variable is carrying some weight in terms of documenting what is going on here — it doesn't seem to be indicated explicitly that the reason for memoizing the attribute handler is purely for performance and isn't otherwise logically required. I mean, I know we have the docstring of `__pydantic_setattr_handlers__`, but that is far away from the implementing code and still doesn't make it clear that the current logic _only_ makes use of the memoization for performance reasons (and isn't otherwise dependent on the memoization), which may be useful context if we refactor this in the future.

I don't really mind the word fast in the name, but I think either way we should add a comment here reiterating that the memoization is purely for performance reasons. And if we do, the word fast probably becomes less useful.

**Review by dmontagu — COMMENTED** [review]
This is great!

Smart idea to memoize most of the checks that are really only dependent on the class and attribute name. I made a couple notes and generally defer to Sydney on specific style nits, but the overall approach seems sensible to me and a good improvement.

Consider the PR approved by me, at least conceptually; I'm just not explicitly approving due to the nit comments maybe meriting some minor changes before merging.

**Comment by sydney-runkle** [comment]
> Consider the PR approved by me, at least conceptually

Fantastic, thanks for the prompt review.

I've give this a nit-picky review this evening, then we can move forward!

This makes me think more about what else we could potentially memoize in the schema gen department...

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done, simplified the name, it wasnt so pretty yet :)

File: pydantic/main.py
Comment (Viicos) [inline pydantic/main.py:173]:
```suggestion
    """`__setattr__` handlers. Memoizing the handlers leads to a dramatic performance improvement in `__setattr__`"""
```

File: pydantic/main.py
Comment (Viicos) [inline pydantic/main.py:900]:
```suggestion
            elif (setattr_handler := self._setattr_handler(name, value)) is not None:
```

**Review by Viicos — COMMENTED** [review]
Thanks, I think this is a smart idea. We might have to worry about the size of the cache for large models (with a lot of fields). If we encounter such issues, we could use proper functions defined once instead of creating lambdas everytime

**Comment by MarkusSintonen** [comment]
> Thanks, I think this is a smart idea. We might have to worry about the size of the cache for large models (with a lot of fields). If we encounter such issues, we could use proper functions defined once instead of creating lambdas everytime

I wouldnt worry about size of it as anyways all the fields are listed in various ClassVars. However if we want to remove the tiny overhead of field name strs we could push the handler fn into eg FieldInfo/ModelPrivateAttr.

> functions defined once

I purposely didnt want to touch the model generation side to not make it anymore heavier than it already is. Because of the mentioned large models it could just do work for no good reason in case fields are not even used like this.

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

**Comment by Viicos** [comment]
Not sure exactly what you mean by _field name strs_ / _didnt want to touch the model generation side_, but I wanted to say is we could do something like this:

```py
HANDLERS = {
    'descriptor': lambda m, val: attribute.__set__(m, val),
    'cached_property': lambda m, val: m.__dict__.__setitem__(name, val),
    ...
}

def _setattr_handler(name: str, value: Any):
    ...
    if hasattr(attribute, '__set__'):
        return HANDLERS['descriptor']
    ...
    elif isinstance(attr, cached_property):
        return HANDLERS['cached_property']
```

So that we don't create a new function every time.

**Comment by MarkusSintonen** [comment]
> but what I wanted to say is we could do something like this

Ah yes I see! That would make sense yes

**Comment by MarkusSintonen** [comment]
@Viicos the suggestion now done here https://github.com/pydantic/pydantic/pull/10868/commits/7dfbe50eb291612503d4dca3cb164254778e0b84

File: pydantic/main.py
Comment (sydney-runkle) [inline pydantic/main.py:921]:
```suggestion
                Returns `None` when memoization is not safe, then the attribute is set directly.
```

File: pydantic/main.py
Comment (sydney-runkle) [inline pydantic/main.py:920]: Can you give a bit more context here for folks who might be diving into this code for the first time on memoization and why it's helpful? Feel free to put that in the main docstring description, thanks!

File: pydantic/main.py
Comment (sydney-runkle) [inline pydantic/main.py:914]: Maybe add a comment here that if `None` is returned, `_object_setattr` was used directly?

**Review by Viicos — APPROVED** [review]
Thanks for adding the simple dict. This looks good to me.

Overall this approach can look a bit weird as calling `__setattr__` (i.e. at the _instance_ level) mutates a class variable that will be valid for every instance. But functionally it makes sense.

File: pydantic/main.py
Comment (sydney-runkle) [inline pydantic/main.py:962]: I'm a bit confused how we got this from the previous logic - don't we need to check `name in self.model_extra`?

File: pydantic/main.py
Comment (sydney-runkle) [inline pydantic/main.py:964]:
```suggestion
```

Seems clear enough

**Review by sydney-runkle — COMMENTED** [review]
Looking good! A few more questions / comments.

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py:962]: The check was redundant, as we know the extra attribute needs to go to the special `__pydantic_extra__` when model class does not define the private attribute. There is no need to check `model_extra` (which just checks `__pydantic_extra__`)

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py:915]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py:921]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

File: pydantic/main.py
Comment (MarkusSintonen) [reply pydantic/main.py]: Done

**Review by sydney-runkle — APPROVED** [review]
Great work, thank you!

**Pull request merged**
